app/components/app_programme_session_table_component.rb

```diff
@@ -6,48 +6,60 @@ def initialize(sessions, programme:)
 
     @sessions = sessions
     @programme = programme
+
+    @patient_sessions =
+      PatientSession
+        .where(session: sessions)
+        .eager_load(:patient)
+        .preload(session: :programmes)
+        .group_by(&:session_id)
+
+    @outcomes =
+      Outcomes.new(patient_sessions: PatientSession.where(session: sessions))
   end
 
   private
 
-  attr_reader :sessions, :programme
+  attr_reader :sessions, :programme, :patient_sessions, :outcomes
 
   def cohort_count(session:)
-    session.patient_sessions.length.to_s
+    patient_sessions.fetch(session.id, []).length
   end
 
   def number_stat(session:)
-    session.patient_sessions.select { yield it }.length.to_s
+    patient_sessions.fetch(session.id, []).select { yield it }.length.to_s
   end
 
   def percentage_stat(session:)
-    total_count = session.patient_sessions.length
+    total_count = patient_sessions.fetch(session.id, []).length
     return nil if total_count.zero?
 
-    count = session.patient_sessions.select { yield it }.length
+    count = patient_sessions.fetch(session.id, []).select { yield it }.length
 
     number_to_percentage(count / total_count.to_f * 100.0, precision: 0)
   end
 
   def no_response_count(session:)
-    number_stat(session:) { it.patient.consent_outcome.no_response?(programme) }
+    number_stat(session:) do
```

Comment:
I'm a bit concerned by the `it` syntax (not just here). It's not apparent what `it` stands for. I had to jump into the method to see what it returns. I think we should use a block variable with a suitable name to make the code more readable (where appropriate).